Introduce Tensor API, Tensor Utilities and compatiblity with ONNX RT

mikepapadim opened the pull request:
#### Description

This pull request introduces a new Tensor API v0.1, providing a comprehensive set of classes and utilities for working with tensors of various data types along with compatibility with ONNX RT Java bindings. 

**The primary changes include:**
- **AbstractTensor Interface**: An abstract base class that defines the common contract for all tensor implementations.
- **Tensor Classes**: A set of concrete tensor classes for different data types, including TensorFloat32, TensorFloat16, TensorInt32, TensorInt16, TensorInt64, TensorFloat64, and TensorByte. These classes extend the AbstractTensor interface and provide type-specific functionality.
- **DType Class**: A utility class that encapsulates data type information for tensors, facilitating type-safe operations and conversions.
- **Shape Class**: A class representing the shape (dimensions) of a tensor.

Also, this PR adds test to show compatibility with ONNX RT Java bindings->

```java
  public void testOnnxCompatibility() throws OrtException {
        Shape shape = new Shape(1, 3, 224, 224);
        TensorFloat32 tornadoTensor = new TensorFloat32(shape);

        tornadoTensor.init(2f);

        OnnxTensor outputTensor = null;

        try (OrtEnvironment env = OrtEnvironment.getEnvironment()) {
            // Load the MobileNet V2 ONNX model
            OrtSession session = env.createSession(MODEL_PATH, new OrtSession.SessionOptions());

            OnnxTensor inputTensor = OnnxTensor.createTensor(env, tornadoTensor.getFloatBuffer(), shape.dimensions());

            Map<String, OnnxTensor> inputMap = new HashMap<>();
            inputMap.put(INPUT_TENSOR_NAME, inputTensor);

            // Run the model inference
            try (OrtSession.Result outputMap = session.run(inputMap)) {
                Optional<OnnxValue> optionalOutputTensor = outputMap.get(OUTPUT_TENSOR_NAME);
                if (optionalOutputTensor.isEmpty()) {
                    throw new IllegalArgumentException("Output tensor not found in model output.");
                }
                outputTensor = (OnnxTensor) optionalOutputTensor.get();

            }
        } finally {
            Assert.assertNotNull(outputTensor);
        }
    }
```

**Overview of the Tensor API Architecture:**
![TensorInt64](https://github.com/beehive-lab/TornadoVM/assets/8652854/d118539d-2f43-4e44-8b03-c0bf05bac137)


**Notes:**
It adds a dependency to ONNX runtime in the `tornado-unittests` packages, as an alternative we could move the compatibility tests into a seperate repo or submodule.


#### Backend/s tested

Mark the backends affected by this PR.

- [X] OpenCL
- [X] PTX
- [ ] SPIRV

#### OS tested

Mark the OS where this PR is tested.

- [X] Linux
- [ ] OSx
- [ ] Windows

#### How to test the new patch?

Provide instructions about how to test the new patch. 

```bash

make jdk21 backends=opencl

# Test how to use Tornado Tensor Types with ONNX RT
tornado-test -V --fast uk.ac.manchester.tornado.unittests.tensors.TestTensorAPIWithOnnx

# Test Tensor Types for all the supported DTypes
tornado-test -V --fast uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes

```

----------------------------------------------------------------------------

hannibalhuang commented:
quick question, is there a document describing the use cases for ONNX RT support for TornadoVM hopes to achieve ?

mikepapadim commented:
> quick question, is there a document describing the use cases for ONNX RT support for TornadoVM hopes to achieve ?

Hello @hannibalhuang. we plan to use for use cases for which data pre-processing is required, such as models used for image classification.

stratika reviewed (commented):
It looks very good. I like the documentation of the new types. I have spotted some minor fixes, and perhaps the addition of a test for `TensorByte` type, which I think is missing.

I tested also on macOS, and I had one test failing:
```bash
Running test: testTensorFloat16Add       ................  [FAILED] 
		\_[REASON] Unable to make field final long jdk.internal.foreign.NativeMemorySegmentImpl.min accessible: module java.base does not "opens jdk.internal.foreign" to module tornado.drivers.opencl
```

jjfumero commented on etc/ml-models/mobilenetv2-7.onnx:
Why is this file in the repo?

mikepapadim replied:
I added a unit-test to import and run inference directly on an onnx model. We could have a smaller or custom model instead, but it need to be in an onnx format.

mikepapadim replied:
This is fixed, now it downlaods the model on the fly and put it into a temporary dir

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/arrays/TornadoNativeArray.java:
```diff
@@ -33,7 +42,10 @@
-public abstract sealed class TornadoNativeArray permits ByteArray, CharArray, DoubleArray, FloatArray, IntArray, LongArray, ShortArray, HalfFloatArray {
+public abstract sealed class TornadoNativeArray //
+        permits ByteArray, CharArray, DoubleArray, FloatArray, HalfFloatArray, //
+        IntArray, LongArray, ShortArray, Tensor, TensorByte, TensorFloat16, //
```
The permits should only include the base class for Tensors. And then the Tensor class itself include the tensor subtypes. Is this possible?

mikepapadim replied:
I did an experiment on that, but it seems that it is not allowed.

I am getting the following error:` Invalid permits clause: 'TensorByte' must directly extend 'TornadoNativeArray'`

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/AbstractTensor.java:
```diff
@@ -0,0 +1,27 @@
+ */
+package uk.ac.manchester.tornado.api.types.tensors;
+
+public interface AbstractTensor {
```
This is an interface, not an abstract class. To me, this should be in the `Tensor` class, rather than a separate interface.

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/DType.java:
```diff
@@ -0,0 +1,116 @@
+import java.lang.foreign.ValueLayout;
+
+/**
+ * The {@code DType} enum represents the various data types supported by a data management or processing system.
```
What do you want to say here? Which data management? Which processing system?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/DType.java:
```diff
@@ -0,0 +1,116 @@
+ * Each enum constant specifies a data type, its size in bytes, and its corresponding {@link ValueLayout} for memory layout.
+ * This allows for easy management and interpretation of different types of data in a uniform way.
+ */
+public enum DType {
```
From the Java side, how do you control is signed or unsigned? How unsigned data types are used? Do you have any example?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Shape.java:
```diff
@@ -0,0 +1,107 @@
+public record Shape(long... dimensions) {
+
+    public Shape(long... dimensions) {
+        this.dimensions = dimensions.clone();
```
Why do you need to clone the input array?

mikepapadim replied:
fixed, this was rendundant

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Shape.java:
```diff
@@ -0,0 +1,107 @@
+    }
+
+    public long[] dimensions() {
+        return dimensions.clone();
```
Same here

mikepapadim replied:
fixed

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Shape.java:
```diff
@@ -0,0 +1,107 @@
+
+    public int getSize() {
+        int size = 1;
+        for (long dim : dimensions) {
```
You can use the Java 8 Stream API to perform this computation.

mikepapadim replied:
fixed

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Shape.java:
```diff
@@ -0,0 +1,107 @@
+        return size;
+    }
+
+    public Shape reshape(long... newDimensions) {
```
Add Javadoc

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,256 @@
+import uk.ac.manchester.tornado.api.types.HalfFloat;
+import uk.ac.manchester.tornado.api.types.arrays.TornadoNativeArray;
+
+@SegmentElementSize(size = 4) //This needs to be fixed
```
In my new, this class does not have a `SegmentElementSize`, but rather other type of annotation that the TornadoVM JIT compiler can take to explore the sub-types of tensor. `Tensor`, in my view, can't be final, thus, other subtypes extends from the `Tensor` class, such as `TensorFloat32`.

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,256 @@
+     *     the float array to convert into a tensor
+     * @return a new tensor containing the provided values
+     */
+    public static Tensor fromArray(float[] values) {
```
There is only a static method to build a tensor from a `float` array. What if the input is an `int[]` array?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,256 @@
+
+    @Override
+    public long getNumBytesOfSegmentWithHeader() {
+        return 0;
```
Why this method returns `0`?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,256 @@
+
+    @Override
+    public long getNumBytesOfSegment() {
+        return 0;
```
Why this method returns `0`?

mikepapadim replied:
fixed

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,256 @@
+    }
+
+    public long getNumBytesWithoutHeader() {
+        return 0;
```
Why this method returns `0`?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,256 @@
+        return sb.toString();
+    }
+
+}
```
After looking at this class. This only works for FP32 and FP16, but then there are the rest of the classes, such as Int, byte, etc., that are built using the specialized classes.

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorByte.java:
```diff
@@ -0,0 +1,130 @@
+import static java.lang.foreign.ValueLayout.JAVA_BYTE;
+import static java.lang.foreign.ValueLayout.JAVA_SHORT;
+
+public final class TensorByte extends TornadoNativeArray implements AbstractTensor {
```
Specialised `Tensor` classes should extend `Tensor` class instead. Any specific reason/s for the current design?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorFloat16.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Override
+    public int getElementSize() {
+        return 0;
```
Why does it return `0`?

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorFloat64.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    public int getElementSize() {
+        return 0;
```
Why this method returns `0`?

mikepapadim replied:
fixed

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorInt16.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    public int getElementSize() {
+        return 0;
```
Why this method returns `0`?

mikepapadim replied:
fixed

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorInt32.java:
```diff
@@ -0,0 +1,138 @@
+
+    @Override
+    public int getElementSize() {
+        return 0;
```
Why this method returns `0`?

mikepapadim replied:
fixed

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorInt64.java:
```diff
@@ -0,0 +1,140 @@
+
+    @Override
+    public int getElementSize() {
+        return 0;
```
Why this method returns `0`?

mikepapadim replied:
fixed

jjfumero commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:
```diff
@@ -0,0 +1,333 @@
+
+        tensorA.init(new HalfFloat(1f));
+        
+        System.out.println("Half-precision tensor:");
+        System.out.println(STR."Shape: \{tensorA.getShape()}");
+        System.out.println(STR."Data type: \{tensorA.getDTypeAsString()}");
+        System.out.println(STR."Shape as TF: \{tensorA.getShape().toTensorFlowShapeString()}");
+        System.out.println(STR."Shape as ONNX: \{tensorA.getShape().toONNXShapeString()}");
```
Test with with assertions instead

mikepapadim replied:
fixed

mikepapadim commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:
```diff
@@ -0,0 +1,333 @@
+        executionPlan.execute();
+
+        for (int i = 0; i < tensorC.getSize(); i++) {
+            //            Assert.assertEquals(tensorC.get(i), (float) HalfFloat.add(tensorA.get(i), tensorB.get(i)), 0.00f);
```
fixed

mikepapadim commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:
```diff
@@ -0,0 +1,333 @@
+import uk.ac.manchester.tornado.api.types.tensors.DType;
+import uk.ac.manchester.tornado.api.types.tensors.Tensor;
+import uk.ac.manchester.tornado.api.types.tensors.Shape;
+import uk.ac.manchester.tornado.api.types.tensors.TensorFloat16;
```
fixed

mikepapadim commented:
This is now ready for another iteration.

1) I refactored the tensors classes to inherit from Tensor class which is only permitted in TornadoNativeArray
2) Renamed fp tensors
3) Download onnx models on the fly
4) add javadocs in shape class methods
5) add factory methods for initiliaze
6) add methods to retun tensor segments as bytebuffers

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,72 @@
+public non-sealed class Tensor extends TornadoNativeArray {
+    @Override
+    public int getSize() {
+        return 0;
```
The implementation of this class is empty. This class should invoke to the child implementation, so probably you need a Tensor type in the constructor and keep a field with the implementation, or build one your own based on the DType. 
I think the second option is the one to go

jjfumero commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorAPIWithOnnx.java:
```diff
@@ -0,0 +1,127 @@
+            }
+        } finally {
+            Assert.assertNotNull(outputTensor);
+            //            cleanupDownloadedModel(modelPath);
```
Remove this line

mikepapadim replied:
done

jjfumero commented:
Is this PR ready for a second review?

mikepapadim commented:
> Is this PR ready for a second review?

Yes

stratika commented:
I have a problem when running the following tests with OpenCL: 
```bash
tornado-test -V --fast uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes
tornado --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes"
WARNING: Using incubator modules: jdk.incubator.vector
[ERROR] clEnqueueReadBuffer, code = -5 n[TornadoVM-OCL-JNI] ERROR : clEnqueueReadBuffer -> Returned: [JNI] uk.ac.manchester.tornado.drivers.opencl> notify error:
[JNI] uk.ac.manchester.tornado.drivers.opencl> -5CL_OUT_OF_RESOURCES error executing CL_COMMAND_READ_BUFFER on NVIDIA RTX A2000 8GB Laptop GPU (Device 0).


[TornadoVM-OCL-JNI] ERROR : clFlush -> Returned: -9999
[JNI] uk.ac.manchester.tornado.drivers.opencl> notify error:
[JNI] uk.ac.manchester.tornado.drivers.opencl> Unknown error executing clFlush on NVIDIA RTX A2000 8GB Laptop GPU (Device 0).

[TornadoVM-OCL-JNI] ERROR : clFinish -> Returned: -36
[TornadoVM-OCL-JNI] ERROR : clEnqueueWriteBuffer -> Returned: -5
[JNI] uk.ac.manchester.tornado.drivers.opencl> notify error:
[JNI] uk.ac.manchester.tornado.drivers.opencl> CL_OUT_OF_RESOURCES error executing CL_COMMAND_WRITE_BUFFER on NVIDIA RTX A2000 8GB Laptop GPU (Device 0).
```

Do they work for you?

jjfumero commented:
No. Same error in my configuration (OpenCL)

```bash
tornado-test -V --fast uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes 
tornado --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes"
WARNING: Using incubator modules: jdk.incubator.vector
[ERROR] clEnqueueReadBuffer, code = -5 n[TornadoVM-OCL-JNI] ERROR : clEnqueueReadBuffer -> Returned: [JNI] uk.ac.manchester.tornado.drivers.opencl> notify error:
[JNI] uk.ac.manchester.tornado.drivers.opencl> -5CL_OUT_OF_RESOURCES error executing CL_COMMAND_READ_BUFFER on NVIDIA GeForce RTX 3070 (Device 0).


[TornadoVM-OCL-JNI] ERROR : clFlush -> Returned: -9999
[JNI] uk.ac.manchester.tornado.drivers.opencl> notify error:
[JNI] uk.ac.manchester.tornado.drivers.opencl> Unknown error executing clFlush on NVIDIA GeForce RTX 3070 (Device 0).

[TornadoVM-OCL-JNI] ERROR : clFinish -> Returned: -36
[TornadoVM-OCL-JNI] ERROR : clEnqueueWriteBuffer -> Returned: [JNI] uk.ac.manchester.tornado.drivers.opencl> notify error:
[JNI] uk.ac.manchester.tornado.drivers.opencl> CL_OUT_OF_RESOURCES error executing CL_COMMAND_WRITE_BUFFER on NVIDIA GeForce RTX 3070 (Device 0).
```

jjfumero commented:
This is related to offsets and sizes when moving data from the host to the device.

mikepapadim commented:
if i run the tests individually, they are passing:

```
tornado-test --printKernel -V --fast uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes#testTensorInt64Add
```

mikepapadim commented:
Now, it should be fine

stratika commented on tornado-drivers/opencl/src/main/java/uk/ac/manchester/tornado/drivers/opencl/runtime/OCLTornadoDevice.java:
```diff
@@ -57,6 +57,7 @@
 import uk.ac.manchester.tornado.api.types.arrays.FloatArray;
 import uk.ac.manchester.tornado.api.types.arrays.HalfFloatArray;
 import uk.ac.manchester.tornado.api.types.arrays.IntArray;
+import uk.ac.manchester.tornado.api.types.tensors.TensorFP16;
```
Remove this, seems to be unused.

stratika commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:
```diff
@@ -0,0 +1,339 @@
+        // Create a tensor to store the result of addition
+        TensorInt64 tensorC = new TensorInt64(shape);
+
+        tensorA.init(20l);
+        tensorB.init(3000l);
+        tensorA.init(0l);
```
By running,

```bash
make checkstyle:
[INFO] There are 5 errors reported by Checkstyle 9.3 with tornado-assembly/src/etc/checkstyle.xml ruleset.
[ERROR] src/main/java/uk/ac/manchester/tornado/unittests/api/TestConcat.java:[128,16] (misc) UpperEll: Should use uppercase 'L'.
[ERROR] src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorAPIWithOnnx.java:[112,28] (misc) ArrayTypeStyle: Array brackets at illegal position.
[ERROR] src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:[283,22] (misc) UpperEll: Should use uppercase 'L'.
[ERROR] src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:[284,22] (misc) UpperEll: Should use uppercase 'L'.
[ERROR] src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:[285,22] (misc) UpperEll: Should use uppercase 'L'.
```

stratika commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:
```diff
@@ -0,0 +1,339 @@
+        // Create a tensor to store the result of addition
+        TensorInt64 tensorC = new TensorInt64(shape);
+
+        tensorA.init(20l);
```
```suggestion
        tensorA.init(20L);
```

stratika commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorTypes.java:
```diff
@@ -0,0 +1,339 @@
+        TensorInt64 tensorC = new TensorInt64(shape);
+
+        tensorA.init(20l);
+        tensorB.init(3000l);
```
```suggestion
        tensorB.init(3000L);
```

stratika commented on tornado-unittests/src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorAPIWithOnnx.java:
```diff
@@ -0,0 +1,121 @@
+        path = modelPath.toString();
+
+        try (BufferedInputStream in = new BufferedInputStream(url.openStream()); FileOutputStream fileOutputStream = new FileOutputStream(modelPath.toFile())) {
+            byte dataBuffer[] = new byte[1024];
```
```bash
src/main/java/uk/ac/manchester/tornado/unittests/tensors/TestTensorAPIWithOnnx.java:[112,28] (misc) ArrayTypeStyle: Array brackets at illegal position.
```
```suggestion
            byte[] dataBuffer = new byte[1024];
```

stratika reviewed (commented):
Added some suggested fixes for style issues in the current code base.

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/Tensor.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    @Override
+    public MemorySegment getSegment() {
```
I still don't get the purpose of the implementation of these methods, Can you set to abstract method and then all subtypes implement these methods using the right fields?

mikepapadim replied:
these methods cannot be abstract because they are overriding the methods from the TornadoNativeArray that Tensor extends

mikepapadim replied:
this is fixed now

jjfumero commented on tornado-api/src/main/java/uk/ac/manchester/tornado/api/types/tensors/TensorFP16.java:
```diff
@@ -0,0 +1,159 @@
+     */
+
+    public TensorFP16(Shape shape) {
+        super(DType.HALF_FLOAT, shape);
```
FP16 is 2 bytes. so it shoult be `SHORT`, correct?

mikepapadim replied:
Indeed, but the DType class has the dtypes in format that are stored in model files

jjfumero commented:
SPIR-V backend:

```
tornado-test -V --fast uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes
tornado --jvm "-Xmx6g -Dtornado.recover.bailout=False -Dtornado.unittests.verbose=True "  -m  tornado.unittests/uk.ac.manchester.tornado.unittests.tools.TornadoTestRunner  --params "uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes"
WARNING: Using incubator modules: jdk.incubator.vector
Test: class uk.ac.manchester.tornado.unittests.tensors.TestTensorTypes
	Running test: testHelloTensorAPI         ................  [PASS] 
	Running test: testTensorFloat16Add       ................  [FAILED] 
		\_[REASON] expected:<0.0> but was:<5.0>
	Running test: testTensorFloat32Add       ................  [FAILED] 
		\_[REASON] expected:<0.0> but was:<3000.0>
	Running test: testTensorFloat64Add       ................  [FP64 UNSUPPORTED FOR CURRENT DEVICE] 
	Running test: testTensorInt16Add         ................  [FAILED] 
		\_[REASON] expected:<0.0> but was:<300.0>
	Running test: testTensorInt32Add         ................  [FAILED] 
		\_[REASON] expected:<0.0> but was:<300.0>
	Running test: testTensorInt64Add         ................  [FAILED] 
		\_[REASON] expected:<0.0> but was:<3000.0>
	Running test: testTensorByte             ................  [FAILED] 
		\_[REASON] expected:<0.0> but was:<44.0>

```

I am using the Intel compute runtime: `24.09.28717.12`

jjfumero commented:
Pending:
- Investigate the failures in the SPIR-V backend.

mikepapadim commented:
Pending: fix for FP16

mikepapadim commented:
> Pending: fix for FP16

this is fixed now
